### PR TITLE
SPARKC-105: Fix setting custom connection factory which was blocked by t...

### PR DESCRIPTION
**Pull request by pkolaczk**
...he new configuration check system. Provided a way to set custom auth config and connection factory properties.

### CODE REVIEW

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/cql/AuthConf.scala [hunk 1]
```diff
@@ -35,7 +35,23 @@ case class PasswordAuthConf(user: String, password: String) extends AuthConf {
...
+
   def authConf(conf: SparkConf): AuthConf
+
+  /** List of extra properties allowed in SparkConf passed to `authConf` method */
+  def properties: Set[String] = Set.empty
+}
+
+object AuthConfFactory {
+  val AuthConfFactoryProperty = "spark.cassandra.auth.conf.factory"
+  val Properties = Set(AuthConfFactoryProperty)
```
Comment (helena) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/cql/AuthConf.scala]: Thank you for one-lining this. 
Shouldn't it be a final val, final val are capitalized

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/cql/CassandraConnectionFactory.scala [hunk 2]
```diff
@@ -81,6 +83,7 @@ object DefaultConnectionFactory extends CassandraConnectionFactory {
   * used when establishing connections to Cassandra. */
 object CassandraConnectionFactory {
   val ConnectionFactoryProperty = "spark.cassandra.connection.factory"
+  val Properties = Set(ConnectionFactoryProperty)
```
Comment (helena) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/cql/CassandraConnectionFactory.scala]: final val ... for both?

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/util/ConfigCheck.scala [hunk 3]
```diff
@@ -29,7 +32,11 @@ object ConfigCheck {
    * @param conf SparkConf object to check
    */
   def checkConfig(conf: SparkConf): Unit = {
-    val unknownProps = getUnknownProperties(conf)
+    val connectionFactory = CassandraConnectionFactory.fromSparkConf(conf)
```
Comment (helena) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/util/ConfigCheck.scala]: These should be an apply but i know you disagree.
Apply's are for creating, the signature of the applys is all one needs, not "fromSparkConf"

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/util/ConfigCheck.scala [hunk 4]
```diff
@@ -40,7 +47,8 @@ object ConfigCheck {
     }
   }
 
-  def getUnknownProperties(conf: SparkConf): Seq[String] = {
+  def getUnknownProperties(conf: SparkConf, extraProps: Set[String] = Set.empty): Seq[String] = {
```
Comment (helena) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/util/ConfigCheck.scala]: who added a 'get*', this is not java

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/util/ConfigCheck.scala [hunk 5]
```diff
@@ -56,9 +64,10 @@ object ConfigCheck {
    *
    * Fuzziness is determined by MatchThreshold
    */
-  def getSuggestedVars(unknownProp: String): Option[Seq[String]] = {
+  def getSuggestedVars(unknownProp: String, extraProps: Set[String] = Set.empty): Option[Seq[String]] = {
```
Comment (helena) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/util/ConfigCheck.scala]: I know you didn't write this, i'm just seeing it for the first time.
It is odd to return an Option of some collection when it should simply be Seq.empty in this case

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/cql/AuthConf.scala (hunk 1 above)
Comment (pkolaczk) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/cql/AuthConf.scala]: What is the point of final val for an object? Objects are final by definition.

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/util/ConfigCheck.scala (hunk 3 above)
Comment (pkolaczk) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/util/ConfigCheck.scala]: Ok, we can address it in another ticket. On the other hand, multiple apply() overload with *default* parameters is not allowed, while naming them differently allows default params.

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/util/ConfigCheck.scala (hunk 5 above)
Comment (pkolaczk) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/util/ConfigCheck.scala]: Oh yes, it caught my eye during review, too, however it turned out there was an *intended* semantic difference. Now when I look at it for the second time, I think it *can* be simplified to Seq[String], it just needs some minor modification of the call-site.

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/util/ConfigCheck.scala (hunk 4 above)
Comment (pkolaczk) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/util/ConfigCheck.scala]: *nod*

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/cql/AuthConf.scala (hunk 1 above)
Comment (helena) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/cql/AuthConf.scala]: true, i am thinking in terms of a standard Akka class of settings which use final val MyVal syntax vs an object

File: spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/util/ConfigCheck.scala (hunk 4 above)
Comment (pkolaczk) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/util/ConfigCheck.scala]: BTW: In defense of Javaism, I can say that having "get" in the beginning makes it somehow handy when naming things on the call site. Compare those two snippets:
```scala
def getUnknownProperties(...) = ???
// later in the code:
val unknownProperties = getUnknownProperties(...)  // obvious
```
```scala
def unknownProperties(...) = ???
val unknownProperties = unknownProperties(...)  // damn; now it requires either explicit type annotation, or a different name
```

Comment (helena) [inline spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/util/ConfigCheck.scala]: Just a naming change, I'd hate to see java beans-isms sneak in

**Comment by helena** [comment]
LGTM

**Pull request merged**